### PR TITLE
Migrate Tag component to Typescript

### PR DESCRIPTION
**Pull request by bradd123**
#### Summary

Migrate type checking of Tag component from `PropTypes` to `Typescript`

### CODE REVIEW

File: packages/components/tag/src/tag.tsx [hunk 1]
```diff
@@ -11,7 +10,73 @@ import AccessibleButton from '@commercetools-uikit/accessible-button';
...
+    | 13
+    | 14
+    | 15
+    | 16
+    | 'scale'
+    | 'auto';
+  /**
+   * Content rendered within the tag
+   */
+  children: ReactNode;
```
Comment (adnasa) [inline packages/components/tag/src/tag.tsx:67]: This is the only prop that is required. The rest are optional.

File: packages/components/tag/src/tag.tsx [hunk 2]
```diff
@@ -11,7 +10,73 @@ import AccessibleButton from '@commercetools-uikit/accessible-button';
...
+  /**
+   * Called when remove button is clicked.
+   * <br />
+   * Signature: `(event) => void`
+   */
+  onRemove: (event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>) => void;
+  /**
+   * Called when tag element is clicked. This is not called when remove button is clicked.
+   * <br />
+   * Signature: `(event) => void`
```
Comment (adnasa) [inline packages/components/tag/src/tag.tsx:39]: Remove this. the signature will be identified by the generator

File: packages/components/tag/src/tag.tsx [hunk 3]
```diff
@@ -11,7 +10,73 @@ import AccessibleButton from '@commercetools-uikit/accessible-button';
...
+   */
+  to: string;
+  /**
+   * Disable the tag element along with the option to remove it.
+   */
+  isDisabled: boolean;
+  /**
+   * Called when remove button is clicked.
+   * <br />
+   * Signature: `(event) => void`
```
Comment (adnasa) [inline packages/components/tag/src/tag.tsx:33]: Can be removed

File: packages/components/tag/src/tag-body.tsx [hunk 4]
```diff
@@ -1,16 +1,40 @@
...
   designTokens,
 } from '@commercetools-uikit/design-system';
 import Text from '@commercetools-uikit/text';
 
-const Body = styled.div``;
+type TTagBodyProps = {
+  as?: ElementType;
+  to?: string | LocationDescriptor;
+  type: string;
+  onClick?: (
```
Comment (bradd123) [inline packages/components/tag/src/tag-body.tsx:21]: There is a conflict when we are using `to` and `onClick`. Fixing one will cause other to throw an error.

- When we are using `MouseEvent<HTMLDivElement>`, `onClick` prop is throwing this error
`Type '((event: MouseEvent<HTMLDivElement, MouseEvent> | KeyboardEvent<HTMLDivElement>) => void) | undefined' is not assignable to type 'MouseEventHandler<HTMLButtonElement> | undefined'.
  Type '(event: MouseEvent<HTMLDivElement, MouseEvent> | KeyboardEvent<HTMLDivElement>) => void' is not assignable to type 'MouseEventHandler<HTMLButtonElement>'.
    Types of parameters 'event' and 'event' are incompatible.
      Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is not assignable to type 'MouseEvent<HTMLDivElement, MouseEvent> | KeyboardEvent<HTMLDivElement>'.
        Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is not assignable to type 'MouseEvent<HTMLDivElement, MouseEvent>'.
          Property 'align' is missing in type 'HTMLButtonElement' but required in type 'HTMLDivElement'.ts(2322)
lib.dom.d.ts(6651, 5): 'align' is declared here.
index.d.ts(1457, 9): The expected type comes from property 'onClick' which is declared here on type 'IntrinsicAttributes & { theme?: Theme | undefined; as?: ElementType<any> | undefined; } & ClassAttributes<HTMLButtonElement> & ButtonHTMLAttributes<...> & { ...; }'` 

- When we are using `MouseEvent<HTMLButtonElement>`, `to` prop is throwing this error

`Type '{ children: Element; to: LocationDescriptor<unknown> | undefined; as: ElementType<any> | undefined; css: (false | SerializedStyles | SerializedStyles[] | undefined)[]; onClick: ((event: MouseEvent<...> | KeyboardEvent<...>) => void) | undefined; }' is not assignable to type 'IntrinsicAttributes & { theme?: Theme | undefined; as?: ElementType<any> | undefined; } & ClassAttributes<HTMLButtonElement> & ButtonHTMLAttributes<...> & { ...; }'.
  Property 'to' does not exist on type 'IntrinsicAttributes & { theme?: Theme | undefined; as?: ElementType<any> | undefined; } & ClassAttributes<HTMLButtonElement> & ButtonHTMLAttributes<...> & { ...; }'`

**Review by adnasa — CHANGES_REQUESTED** [review]
Welcome to the ui-kit 🙌🏽

File: packages/components/tag/src/tag-body.tsx [hunk 5]
```diff
@@ -1,16 +1,40 @@
...
+  ) => void;
+  onRemove: (
+    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>
+  ) => void;
+  isDisabled: boolean;
+  children: ReactNode;
+  styles?: Record<string, SerializedStyles>;
+};
+
+const Body = styled.button``;
```
Comment (bradd123) [inline packages/components/tag/src/tag-body.tsx:32]: Here I have tried with various other elements like `styled.a` `styled.link`. But they are also having the same problem.

Comment (adnasa) [reply packages/components/tag/src/tag-body.tsx:32]: What were you trying to achieve?

File: packages/components/tag/src/tag-body.tsx [hunk 6]
```diff
@@ -1,16 +1,40 @@
...
+import React, {
+  MouseEvent,
+  KeyboardEvent,
+  ReactNode,
+  ElementType,
+} from 'react';
 import styled from '@emotion/styled';
-import { css, useTheme } from '@emotion/react';
+import type { Theme } from '@emotion/react';
+import type { LocationDescriptor } from 'history';
```
Comment (adnasa) [inline packages/components/tag/src/tag-body.tsx:9]: move these to the top of the file

File: packages/components/tag/src/tag-body.tsx (hunk 5 above)
Comment (bradd123) [reply packages/components/tag/src/tag-body.tsx:32]: We are trying to render the component as `react-router-dom`'s `Link` if there is an url as `to` property and it's not disabled. We are also having `onClick` property.

The problem is `Property 'to' does not exist on any element`. I have gone through types rabbit hole and found out that none of the `HTMLAnchorElement`, `HTMLButtonElement`, etc has a key called `to`. So those are rejecting this prop.

**Review by mohib0306 — APPROVED** [review]
LGTM

File: packages/components/tag/src/tag.tsx [hunk 7]
```diff
@@ -11,9 +9,73 @@ import AccessibleButton from '@commercetools-uikit/accessible-button';
...
 import TagBody from './tag-body';
 
-const Tag = (props) => {
-  const linkProps =
-    props.to && !props.isDisabled ? { as: Link, to: props.to } : {};
+type TTagProps = {
+  /**
+   * Indicates color scheme of the tag.
+   */
+  type: 'normal' | 'warning';
```
Comment (adnasa) [inline packages/components/tag/src/tag.tsx:16]: type is not required

File: packages/components/tag/src/tag.tsx [hunk 8]
```diff
@@ -11,9 +9,73 @@ import AccessibleButton from '@commercetools-uikit/accessible-button';
...
+  /**
+   * Called when tag element is clicked. This is not called when remove button is clicked.
+   */
+  onClick?: (
+    event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>
+  ) => void;
+  /**
+   * Horizontal size limit of the input field.
+   */
+  horizontalConstraint:
```
Comment (adnasa) [inline packages/components/tag/src/tag.tsx:44]: horizontalConstraint is not required

File: packages/components/tag/src/tag-body.tsx [hunk 9]
```diff
@@ -72,12 +95,14 @@ const getContentWrapperStyles = (props, theme) => {
   `;
 };
 
-const TagBody = (props) => {
-  const theme = useTheme();
+const TagBody = (props: TTagBodyProps) => {
+  const theme: Theme = useTheme();
+  const linkProps =
+    props.to && !props.isDisabled ? { as: Link, to: props.to } : {};
```
Comment (adnasa) [inline packages/components/tag/src/tag-body.tsx:101]: this is changing implementation.
why did we move this from `Tag`?

File: packages/components/tag/src/tag-body.tsx [hunk 10]
```diff
@@ -1,16 +1,36 @@
...
+type TTagBodyProps = {
+  to?: string | LocationDescriptor;
+  type: string;
+  onClick?: (
+    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>
+  ) => void;
+  onRemove?: (
+    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>
+  ) => void;
+  isDisabled?: boolean;
```
Comment (adnasa) [inline packages/components/tag/src/tag-body.tsx:23]: required

**Review by adnasa — CHANGES_REQUESTED** [review]

**Comment by bradd123** [comment]
Thanks @mohib0306 and @adnasa for helping me to resolve some of the type issues.

File: packages/components/tag/src/tag-body.tsx [hunk 11]
```diff
@@ -1,34 +1,32 @@
...
-  type: string;
-  onClick?: (
-    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>
-  ) => void;
-  onRemove?: (
-    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>
-  ) => void;
+  to?: TTagProps['to'];
+  as?: ElementType;
+  type?: string;
```
Comment (adnasa) [inline packages/components/tag/src/tag-body.tsx:16]: hey why not `TTagProps['type']` here?

Comment (bradd123) [reply packages/components/tag/src/tag-body.tsx:16]: The existing code has different types in `tag` and `tag-body`. 

But on second thought, having `'normal' | 'warning` in both places makes sense.

**Review by adnasa — APPROVED** [review]

**Comment by bradd123** [comment]
Manually tested the Tag on staging environment. Everything is working as expected.

**Pull request merged**